File: PaperPlugin/src/main/java/net/onelitefeather/clipboardconnect/paper/ClipboardConnect.java
package net.onelitefeather.clipboardconnect.paper;

import com.fastasyncworldedit.core.util.TaskManager;
import com.sk89q.worldedit.WorldEdit;
import com.sk89q.worldedit.extension.platform.Actor;
import net.kyori.adventure.key.Key;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.translation.GlobalTranslator;
import net.kyori.adventure.translation.TranslationRegistry;
import net.kyori.adventure.util.UTF8ResourceBundleControl;
import net.onelitefeather.clipboardconnect.api.config.PluginConfig;
import net.onelitefeather.clipboardconnect.api.config.PluginConfigProvider;
import net.onelitefeather.clipboardconnect.api.event.ClipboardReadyEvent;
import net.onelitefeather.clipboardconnect.api.transfer.TransferStrategy;
import net.onelitefeather.clipboardconnect.paper.commands.CommandRegistry;
import net.onelitefeather.clipboardconnect.paper.transfer.TransferService;
import net.onelitefeather.clipboardconnect.paper.translations.PluginTranslationRegistry;
import net.onelitefeather.clipboardconnect.strategies.RedisTransferStrategy;
import net.onelitefeather.clipboardconnect.strategies.ZstdRedisTransferStrategy;
import org.bstats.bukkit.Metrics;
import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;
import org.redisson.Redisson;
import org.redisson.api.RedissonClient;
import org.redisson.config.Config;

import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Locale;
import java.util.Objects;
import java.util.ResourceBundle;
import java.util.concurrent.Executor;

public final class ClipboardConnect extends JavaPlugin {


    public static final Component PREFIX = MiniMessage.miniMessage().deserialize("<gold>[<green>ClipboardConnect<gold>] ");
    public static final boolean isFastAsyncWorldEdit = Bukkit.getPluginManager().isPluginEnabled("FastAsyncWorldEdit");
    private PluginConfig pluginConfig;
    private TransferStrategy transferStrategy;
    private TransferService transferService;
    @Override
    public void onLoad() {
        final var dataFolder = this.getDataFolder().toPath();
        if (Files.notExists(dataFolder)) {
            try {
                Files.createDirectories(dataFolder);
            } catch (final Exception e) {
                getSLF4JLogger().error("Failed to create data folder", e);
                Bukkit.getPluginManager().disablePlugin(this);
                return;
            }
        }
        final PluginConfigProvider pluginConfigProvider = PluginConfigProvider.create(this.getDataFolder().toPath());
        if (this.pluginConfig == null) {
            this.pluginConfig = pluginConfigProvider.getPluginConfig();
        }
        if (Files.notExists(dataFolder.resolve(PluginConfig.CONFIG_FILE_NAME))) {
            pluginConfigProvider.saveConfig(pluginConfig);
        }
        loadStrategy();
    }

    private RedissonClient buildRedisson() {
        final var redisFile = this.getDataFolder().toPath().resolve("redis.yml");
        if (Files.notExists(redisFile)) {
            try {
                Files.createFile(redisFile);
            } catch (final Exception e) {
                getSLF4JLogger().error("Failed to create redis file", e);
                Bukkit.getPluginManager().disablePlugin(this);
            }
        }
        try {
            final var config = Config.fromYAML(redisFile.toFile());
            return Redisson.create(config);
        } catch (final IOException e) {
            getSLF4JLogger().error("Failed to load redis file", e);
            Bukkit.getPluginManager().disablePlugin(this);
        } catch (final Exception e) {
            getSLF4JLogger().error("Failed to create redis client", e);
            Bukkit.getPluginManager().disablePlugin(this);
        }
        throw new UnsupportedOperationException("The redis file had some issues");
    }

    @Override
    public void onEnable() {
        new Metrics(this, 20460);
        registerTranslations();
        this.transferService = new TransferService(this);
        var commandRegistry = new CommandRegistry(this);
        commandRegistry.registerCommands();
    }

    private void registerTranslations() {
        final TranslationRegistry translationRegistry = new PluginTranslationRegistry(
                TranslationRegistry.create(
                        Key.key(
                                "clipboardconnect",
                                "translations"
                        )
                )
        );
        translationRegistry.defaultLocale(Locale.US);
        final Path langFolder = getDataFolder().toPath().resolve("lang");
        if (Files.exists(langFolder)) {
            try {
                try (URLClassLoader urlClassLoader = new URLClassLoader(new URL[]{langFolder.toUri().toURL()})) {
                    Locale.availableLocales().map(locale -> {
                        try {
                            return ResourceBundle.getBundle(
                                    "clipboardconnect",
                                    locale,
                                    urlClassLoader,
                                    UTF8ResourceBundleControl.get()
                            );
                        } catch (Exception e) {
                            return null;
                        }
                    }).filter(Objects::nonNull).forEach(bundle -> {
                        try {
                            translationRegistry.registerAll(bundle.getLocale(), bundle, false);
                        } catch (Exception e) {
                            getSLF4JLogger().error("Failed to register translations for {} because: {}", bundle.getLocale(), e.getMessage());
                            getSLF4JLogger().debug("Failed to register translations", e);
                        }
                    });
                }
            } catch (IOException e) {
                getSLF4JLogger().error("Failed to load translations", e);
            }
        } else {
            Locale.availableLocales().map(locale -> {
                try {
                    return ResourceBundle.getBundle(
                            "clipboardconnect",
                            locale,
                            UTF8ResourceBundleControl.get()
                    );
                } catch (Exception e) {
                    return null;
                }
            }).filter(Objects::nonNull).forEach(bundle -> {
                try {
                    translationRegistry.registerAll(bundle.getLocale(), bundle, false);
                } catch (Exception e) {
                    getSLF4JLogger().error("Failed to register translations for {} because: {}", bundle.getLocale(), e.getMessage());
                    getSLF4JLogger().debug("Failed to register translations", e);
                }
            });
        }
        GlobalTranslator.translator().addSource(translationRegistry);
    }

    @Override
    public void reloadConfig() {
        final var dataFolder = this.getDataFolder().toPath();
        if (Files.notExists(dataFolder)) {
            try {
                Files.createDirectories(dataFolder);
            } catch (final Exception e) {
                getSLF4JLogger().error("Failed to create data folder", e);
                Bukkit.getPluginManager().disablePlugin(this);
                return;
            }
        }
        final PluginConfigProvider pluginConfigProvider = PluginConfigProvider.create(this.getDataFolder().toPath());
        this.pluginConfig = pluginConfigProvider.getPluginConfig();
        if (Files.notExists(dataFolder.resolve(PluginConfig.CONFIG_FILE_NAME))) {
            pluginConfigProvider.saveConfig(pluginConfig);
        }
    }

    private void loadStrategy() {
        getSLF4JLogger().debug("Loading strategy model {}", pluginConfig.strategyModel());
        this.transferStrategy = switch (pluginConfig.strategyModel()) {
            case REDIS: yield new RedisTransferStrategy(buildRedisson(), this.pluginConfig, this::clipboardReadyEventCall);
            case ZSTD_REDIS: yield new ZstdRedisTransferStrategy(buildRedisson(), this.pluginConfig, this::clipboardReadyEventCall);
        };
        getSLF4JLogger().info("Loaded strategy model {}", pluginConfig.strategyModel());
    }

    private void clipboardReadyEventCall(Actor actor) {
        WorldEdit.getInstance().getEventBus().post(ClipboardReadyEvent.create(actor));
    }

    public Executor getExecutorService() {
        if (isFastAsyncWorldEdit) {
            return TaskManager.taskManager().getPublicForkJoinPool();
        } else {
            return Bukkit.getScheduler().getMainThreadExecutor(this);
        }
    }

    public TransferService getTransferService() {
        return transferService;
    }

    public TransferStrategy getTransferStrategy() {
        return transferStrategy;
    }

    public PluginConfig getPluginConfig() {
        return pluginConfig;
    }
}

File: src/main/kotlin/net/onelitefeather/clipboardconnect/ClipboardConnect.kt
package net.onelitefeather.clipboardconnect

import cloud.commandframework.annotations.AnnotationParser
import cloud.commandframework.bukkit.CloudBukkitCapabilities
import cloud.commandframework.execution.CommandExecutionCoordinator
import cloud.commandframework.paper.PaperCommandManager
import dev.derklaro.aerogel.Element
import dev.derklaro.aerogel.Injector
import dev.derklaro.aerogel.Order
import dev.derklaro.aerogel.binding.BindingBuilder
import dev.derklaro.aerogel.member.InjectionSetting
import dev.derklaro.aerogel.util.Qualifiers
import dev.derklaro.aerogel.util.Scopes
import jakarta.inject.Inject
import jakarta.inject.Singleton
import net.kyori.adventure.key.Key
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.translation.GlobalTranslator
import net.kyori.adventure.translation.TranslationRegistry
import net.kyori.adventure.util.UTF8ResourceBundleControl
import net.onelitefeather.clipboardconnect.command.ClipboardPlayer
import net.onelitefeather.clipboardconnect.command.ClipboardSender
import net.onelitefeather.clipboardconnect.commands.LoadCommand
import net.onelitefeather.clipboardconnect.commands.SaveCommand
import net.onelitefeather.clipboardconnect.commands.SetupCommand
import net.onelitefeather.clipboardconnect.conversation.ConversationContext
import net.onelitefeather.clipboardconnect.listener.PlayerJoinListener
import net.onelitefeather.clipboardconnect.listener.PlayerQuitListener
import net.onelitefeather.clipboardconnect.listener.SetupListener
import net.onelitefeather.clipboardconnect.services.SetupService
import net.onelitefeather.clipboardconnect.translations.PluginTranslationRegistry
import net.onelitefeather.clipboardconnect.utils.RawTypeMatcher
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import java.io.IOException
import java.net.URLClassLoader
import java.nio.file.Files
import java.util.*
import kotlin.io.path.Path

/**
 * This class represents a ClipboardConnect plugin that provides functionality
 * for managing the clipboard and setup process. It extends the JavaPlugin class
 * and implements various methods for plugin initialization and configuration.
 *
 */
@Singleton
class ClipboardConnect : JavaPlugin() {

    private val prefixComponent = MiniMessage.miniMessage().deserialize("<gold>[<green>ClipboardConnect<gold>] ")

    @Inject
    private lateinit var setupService: SetupService

    override fun onLoad() {
        saveDefaultConfig()
    }

    override fun onEnable() {
        val injector = Injector.newInjector()

        injector.install(BindingBuilder.create()
            .bind(FileConfiguration::class.java)
            .toInstance(config))
            .install(BindingBuilder.create()
            .bind(Element.forType(Component::class.java).requireAnnotation(Qualifiers.named("prefix"))).toInstance(prefixComponent))
            .install(BindingBuilder.create().bind(Element
                .forType(Boolean::class.java)
                .requireAnnotation(Qualifiers.named("fawe"))
            ).toInstance(Bukkit.getPluginManager().getPlugin("FastAsyncWorldEdit") != null))
            .install(BindingBuilder.create()
                .bindAll(ClipboardConnect::class.java)
                .scoped(Scopes.SINGLETON)
                .toInstance(this))
        injector.memberInjector(ClipboardConnect::class.java).inject(injector.instance(ClipboardConnect::class.java), InjectionSetting.toFlag(
            InjectionSetting.PRIVATE_METHODS,
            InjectionSetting.STATIC_METHODS,
            InjectionSetting.INHERITED_METHODS,
            InjectionSetting.PRIVATE_FIELDS,
            InjectionSetting.STATIC_FIELDS,
            InjectionSetting.INSTANCE_FIELDS,
            InjectionSetting.INHERITED_FIELDS,
            InjectionSetting.RUN_POST_CONSTRUCT_LISTENERS
        ))
    }

    @Inject
    private fun registerTranslations() {
        val translationRegistry: TranslationRegistry = PluginTranslationRegistry(
            TranslationRegistry.create(
                Key.key(
                    "clipboardconnect",
                    "translations"
                )
            )
        )
        translationRegistry.defaultLocale(Locale.US)
        val langFolder = dataFolder.toPath().resolve("lang")
        if (Files.exists(langFolder)) {
            try {
                URLClassLoader(arrayOf(langFolder.toUri().toURL())).use { urlClassLoader ->
                    config.getStringList("translations").stream().map { languageTag: String ->
                        Locale.forLanguageTag(
                            languageTag
                        )
                    }.forEach { r: Locale ->
                        val bundle = ResourceBundle.getBundle(
                            "clipboardconnect",
                            r,
                            urlClassLoader,
                            UTF8ResourceBundleControl.get()
                        )
                        translationRegistry.registerAll(r, bundle, false)
                    }
                }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        } else {
            config.getStringList("translations").stream().map { languageTag: String ->
                Locale.forLanguageTag(
                    languageTag
                )
            }.forEach { r: Locale ->
                val bundle =
                    ResourceBundle.getBundle("clipboardconnect", r, UTF8ResourceBundleControl.get())
                translationRegistry.registerAll(r, bundle, false)
            }
        }
        GlobalTranslator.translator().addSource(translationRegistry)
    }

    @Inject
    private fun bsStats() {
        Metrics(this, 20460)
    }

    /**
     * Initializes the command manager for the ClipboardConnect plugin.
     *
     * @param injector The Injector instance used for dependency injection.
     */
    @Inject
    @Order(10)
    private fun commandManager(injector: Injector) {
        val commandManager = PaperCommandManager(this, CommandExecutionCoordinator.simpleCoordinator(), this::commandSenderChoose, ClipboardSender::getCommandSender)
        if (commandManager.hasCapability(CloudBukkitCapabilities.ASYNCHRONOUS_COMPLETION)) {
            commandManager.registerAsynchronousCompletions()
        }

        if (commandManager.hasCapability(CloudBukkitCapabilities.BRIGADIER)) {
            commandManager.registerBrigadier()
        }
        injector.install(BindingBuilder.create().bind(Element.forType(commandManager.javaClass)).bindMatching(RawTypeMatcher.create(commandManager.javaClass)).toInstance(commandManager))
    }

    /**
     * Parses the given command annotations and creates an instance of AnnotationParser.
     *
     * @param commandManager The PaperCommandManager instance used for command registration.
     * @param injector The Injector instance used for dependency injection.
     */
    @Inject
    @Order(100)
    private fun annotationParser(commandManager: PaperCommandManager<ClipboardSender>, injector: Injector) {
        val annotationParser = AnnotationParser(
            commandManager,
            ClipboardSender::class.java
        )
        injector.install(BindingBuilder.create().bind(Element.forType(annotationParser.javaClass)).bindMatching(RawTypeMatcher.create(annotationParser.javaClass)).toInstance(annotationParser))
    }

    /**
     * Registers event listeners and parses command annotations.
     *
     * @param injector The Injector instance used for dependency injection.
     */
    @Inject
    @Order(150)
    private fun register(injector: Injector) {
        server.pluginManager.registerEvents(injector.instance(SetupListener::class.java), this)
        injector.instance(AnnotationParser::class.java).parse(injector.instance(SetupCommand::class.java))
        val redisFile = Path(dataFolder.toString(), "redis.yml")
        if (Files.exists(redisFile)) {
            server.pluginManager.registerEvents(injector.instance(PlayerQuitListener::class.java), this)
            server.pluginManager.registerEvents(injector.instance(PlayerJoinListener::class.java), this)
            injector.instance(AnnotationParser::class.java).parse(injector.instance(SaveCommand::class.java))
            injector.instance(AnnotationParser::class.java).parse(injector.instance(LoadCommand::class.java))
        } else {
            componentLogger.info(MiniMessage.miniMessage().deserialize("<green>Please run \"/clipboardconnect setup\" ingame"))
        }
    }


    private fun commandSenderChoose(sender: CommandSender): ClipboardSender {
        return if(sender is Player) {
            ClipboardPlayer(sender)
        } else {
            ClipboardSender(sender)
        }
    }

    /**
     * Generates a configuration file based on the provided conversation context.
     *
     * @param conversationContext The conversation context containing the necessary data.
     */
    fun generateConfig(conversationContext: ConversationContext) {
        setupService.generateConfig(conversationContext)
    }
}